### PR TITLE
[Model & Dataset] facebook & sp2gcl

### PR DESCRIPTION
**Pull request by 777sssa**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: gammagl/datasets/facebook.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    def download(self) -> None:
+        download_url(self.url, self.raw_dir)
+
+    def process(self) -> None:
+        data = np.load(self.raw_paths[0], 'r', allow_pickle=True)
+        x = tlx.convert_to_tensor(data['features'], dtype=tlx.float32)
+        y = tlx.convert_to_tensor(data['target'], dtype=tlx.int64)
+        edge_index = tlx.convert_to_tensor(data['edges'], dtype=tlx.int64)
+        edge_index = edge_index.T
```
Comment (dddg617) [inline gammagl/datasets/facebook.py:57]: Have you tried if this can work in the other backend like 'mindspore'?

File: gammagl/datasets/facebook.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            being saved to disk. (default: :obj:`None`)
+        force_reload (bool, optional): Whether to re-process the dataset.
+            (default: :obj:`False`)
+    """
+
+    url = 'https://graphmining.ai/datasets/ptg/facebook.npz'
+
+    def __init__(
+        self,
+        root: str,
```
Comment (dddg617) [inline gammagl/datasets/facebook.py:33]: Currently, this argument can be optional, as we have a cached mechanism compared to ``PyG``.

File: examples/sp2_gcl/node_main.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    def forward(self, data, label):
+        h_node_spa, h_node_spe = self.backbone_network(data['x'], data['edge_index'], data['e'], data['u'])
+        h1 = tlx.l2_normalize(h_node_spa, axis=-1, eps=1e-12)
+        h2 = tlx.l2_normalize(h_node_spe, axis=-1, eps=1e-12)
+        logits = tlx.matmul(h1, h2.transpose(-2, -1)) / self.temp
+        labels = tlx.arange(start=0, limit=h1.shape[0], delta=1, dtype=tlx.int64)
+        loss = 0.5 * tlx.losses.softmax_cross_entropy_with_logits(logits, labels) + 0.5 * tlx.losses.softmax_cross_entropy_with_logits(logits.transpose(-2, -1), labels)
+        return loss
+def main(args):
+    global edge, e, u, test_idx
```
Comment (dddg617) [inline examples/sp2_gcl/node_main.py:66]: What is this line doing?

File: examples/sp2_gcl/node_main.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            if len(data.train_mask.size()) > 1:
+                train_idx = tlx.where(data.train_mask[:, args.seed])[0]
+                val_idx = tlx.where(data.val_mask[:, args.seed])[0]
+                test_idx = tlx.where(data.test_mask)[0]
+            else:
+                train_idx = tlx.where(data.train_mask)[0]
+                val_idx = tlx.where(data.val_mask)[0]
+                test_idx = tlx.where(data.test_mask)[0]
+        else:
+            train_idx, val_idx, test_idx = split(y)
```
Comment (dddg617) [inline examples/sp2_gcl/node_main.py:88]: I do not think this is useful. Usually, the train, valid, test split should be done in the dataset. You may directly use `data.train_mask` .etc to get the idx instead of add a new function in the `util`.

File: gammagl/models/sp2gcl.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
+import tensorlayerx as tlx
+import tensorlayerx.nn as nn
+from gammagl.layers.conv import GCNConv
+class Encoder(nn.Module):
```
Comment (dddg617) [inline gammagl/models/sp2gcl.py:4]: I remember there are many `Encoder` in the gammagl. I do not recommend you to place this function here.

File: gammagl/models/sp2gcl.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    def forward(self, x, edge_index, edge_weight=None):
+        x = self.conv1(x, edge_index, edge_weight=edge_weight)
+        x = self.prelu1(self.bn1(x))
+        x = self.conv2(x, edge_index, edge_weight=edge_weight)
+        x = self.prelu2(self.bn2(x))
+        return x
+
+
+class MLP(nn.Module):
```
Comment (dddg617) [inline gammagl/models/sp2gcl.py:23]: Ditto.

File: gammagl/models/sp2gcl.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    def forward(self, x):
+        x = self.layer1(x)
+        x = self.act_fn(x)
+        x = self.layer2(x)
+        return x
+
+
+
+class EigenMLP(nn.Module):
```
Comment (dddg617) [inline gammagl/models/sp2gcl.py:38]: Ditto.

File: gammagl/utils/split.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import tensorlayerx as tlx
+import numpy as np
+
+def split(y):
```
Comment (dddg617) [inline gammagl/utils/split.py:4]: This util is useless, remove it.

Comment (gyzhou2000) [reply gammagl/utils/split.py:4]: `from sklearn.model_selection import train_test_split`

File: examples/sp2_gcl/readme.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
+@ -0,0 +1,40 @@
+# Graph Contrastive Learning with Stable and Scalable
+
+- Paper link: [https://proceedings.neurips.cc/paper_files/paper/2023/file/8e9a6582caa59fda0302349702965171-Paper-Conference.pdf](https://arxiv.org/abs/2201.11349)
```
Comment (gyzhou2000) [inline examples/sp2_gcl/readme.md:4]: 链接不对

File: gammagl/models/sp2gcl.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+    def forward(self, e, u):
+
+        e = e * 100
+        period_term = tlx.arange(0, self.period)
+        # period_term = period_term.to(u.device)
+        period_e = e.unsqueeze(1) * tlx.pow(2, period_term)
+        # period_e = period_e.to(u.device)
+        fourier_e = tlx.concat([tlx.sin(period_e), tlx.cos(period_e)], axis=-1)
+        h = u @ fourier_e
```
Comment (gyzhou2000) [inline gammagl/models/sp2gcl.py:42]: tlx.unsqueeze，矩阵乘用tlx.matmul

File: examples/sp2_gcl/readme.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+```
+
+
+# Dataset Statics
+
+| Dataset  | Paper Code | Out(th)    |
+|----------|------|------------|
+| PubMed     | 82.3±0.3 | OOM        |
+| Wiki-CS | 79.42±0.19 | 76.79 ± 0.61 |
+| Facebook   | 90.43±0.13 | 85.35±0.26 |
```
Comment (gyzhou2000) [inline examples/sp2_gcl/readme.md:34]: 调参

File: gammagl/models/sp2gcl.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+
+
+class EigenMLP(nn.Module):
+    def __init__(self, input_dim, hidden_dim, output_dim, period):
+        super(EigenMLP, self).__init__()
+        self.k = input_dim
+        self.period = period
+        self.phi = nn.Sequential(nn.Linear(1, 16), nn.ReLU(), nn.Linear(16, 16))
+        self.psi = nn.Sequential(nn.Linear(16, 16), nn.ReLU(), nn.Linear(16, 1))
```
Comment (gyzhou2000) [inline gammagl/models/sp2gcl.py:30]: delete

File: tests/datasets/test_facebook.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+from gammagl.datasets import FacebookPagePage
+
+
+def test():
+	dataset = FacebookPagePage(root='data/facebook')
+	g = dataset[0]
+	pass
```
Comment (gyzhou2000) [inline tests/datasets/test_facebook.py:7]: 节点数量，边数量，节点特征维度，类别数量都判断一下

File: examples/sp2_gcl/sp2gcl_trainer.py [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+            spa_emb = tlx.detach(model.spa_encoder(x, edge))
+            spe_emb = tlx.detach(model.spe_encoder(e, u))
+            acc, pred = node_evaluation((spa_emb + spe_emb)/2, y, train_idx, val_idx, test_idx)
+            print(f'Epoch {epoch+1}/{args.num_epochs}, Accuracy: {acc}')
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--seed', type=int, default=0)
+    parser.add_argument('--cuda', type=int, default=3)
```
Comment (gyzhou2000) [inline examples/sp2_gcl/sp2gcl_trainer.py:134]: seed去点，设置device参考其他trainer写法

File: examples/sp2_gcl/sp2gcl_trainer.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    deg_inv_sqrt = 1.0 / np.sqrt(degree)
+    deg_inv_sqrt[np.isinf(deg_inv_sqrt)] = 0
+    I = csr_matrix(np.eye(num_nodes))
+    D_inv_sqrt = csr_matrix((deg_inv_sqrt, (np.arange(num_nodes), np.arange(num_nodes))))
+    L = I - D_inv_sqrt.dot(data_adj).dot(D_inv_sqrt)
+    e, u = scipy.sparse.linalg.eigsh(L, k=100, which='SM', tol=1e-3)
+    data.e = tlx.convert_to_tensor(e, dtype=tlx.float32)
+    data.u = tlx.convert_to_tensor(u, dtype=tlx.float32)
+
+    return data
```
Comment (gyzhou2000) [inline examples/sp2_gcl/sp2gcl_trainer.py:58]: 试试用 `get_laplacian` 接口替换
